### DISCUSSION TITLE
RabbitMQ Memory Optimization Help

### DISCUSSION THREAD

**Discussion by dormanze** [rabbitmq/rabbitmq-server#9918] [Questions] [answered]
Hello, everybody.
rabbitmq version: 3.11.9
erlang version: 25.22
I have two RabbitMQ instances, which carry about 65000 queues, among which about 5,000 are persistent queues, and others are automatically deleted exclusive queues. The size of a single message is less than 5 KB. The following figure shows the memory usage.

![image](https://github.com/rabbitmq/rabbitmq-server/assets/37854724/62c66144-a8e4-4e04-882a-639d67fa5b52)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/37854724/0b130c49-7f33-4877-9f0f-2c961b619f21)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/37854724/1c12eeb6-5de0-48f3-8277-0230096cb1a0)

I have the following questions:
1:Why does the plug occupy 4.3 GB memory? I have only enabled rabbitmq_web_dispatch, rabbitmq_management, and rabbitmq_management_agent. Why do these plug-ins occupy so much memory?
2:I also found that after I restarted the two instances, only the queue and the consumer was initialized memory was not high. The instance with the highest memory usage was about 6 GB. After a period of message production and consumption, the memory keeps rising until it reaches the current value.I'm a little suspicious of a memory leak.
3:Based on the current usage, do you have any suggestions on modifying the configuration to optimize the current memory usage?

**Comment by mkuratczyk** [comment, marked as answer]
Since you have a lot of queues and channels, the plugin that uses a lot of memory is almost certainly the management plugin, since it collects all those metrics that you are showing in the screenshot. We recommend, especially in environments with a lot of queues/connections/channels, using the Prometheus endpoint for monitoring and turning off the Management metrics. This way you don't store metrics within RabbitMQ, but externally (but you also won't get these graphs in the Management UI). If you are not ready for that, you can play with https://www.rabbitmq.com/management.html#sample-retention to limit the amount of data collected/stored.

Second point is the number of channels you have. Given you have ~300 connections, I'm don't understand why you need ~14000 channels. Perhaps instead of that many channels that declare that many queues, you can just have fewer queues bound with multiple routing keys so that the same apps receive the same messages, but using much fewer channels and queues? That should lower the memory usage significantly.

**Reply by dormanze** [reply]
REST APIs are used to collect some metrics. However, we do not need to retain the cache information of metrics for a long time. Can there be some configurations that allow us to retain the REST API capability without occupying a lot of memory? I have checked the configurations you sent. We have not enabled these configurations yet. Do these configurations have default values?

**Reply by mkuratczyk** [reply]
Yes, the retention settings are exactly for that - granularity and retention period for the stats.
Once again though - the Prometheus endpoint is recommended for collecting data.

**Reply by dormanze** [reply]
Thank you for your answer. I'd like to reconfirm that the example given in the documentation is the default value?
![image](https://github.com/rabbitmq/rabbitmq-server/assets/37854724/9dc328f7-1047-42bf-85b3-f94cfb4391ad)

**Reply by mkuratczyk** [reply]
You can run `rabbitmqctl environment` to see the current value

**Reply by dormanze** [reply]
I checked the configuration of my environment and found that the default values of these parameters were not set.
before i set:
```
 {rabbitmq_management,
     [{content_security_policy,
          "script-src 'self' 'unsafe-eval' 'unsafe-inline'; object-src 'self'"},
      {cors_allow_origins,[]},
      {cors_max_age,1800},
      {http_log_dir,none},
      {listener,[{port,15673},{ip,"172.18.0.41"},{ssl,true}]},
      {load_definitions,none},
      {management_db_cache_multiplier,5},
      {max_http_body_size,10000000},
      {process_stats_gc_timeout,300000},
      {stats_event_max_backlog,250}]},
 {rabbitmq_management_agent,
     [{rates_mode,basic},
      {sample_retention_policies,
          [{global,[{605,5},{3660,60},{29400,600},{86400,1800}]},
           {basic,[{605,5},{3600,60}]},
           {detailed,[{605,5}]}]}]},
```
after i set:
```
 {rabbitmq_management,
     [{content_security_policy,
          "script-src 'self' 'unsafe-eval' 'unsafe-inline'; object-src 'self'"},
      {cors_allow_origins,[]},
      {cors_max_age,1800},
      {http_log_dir,none},
      {listener,[{port,15673},{ip,"172.18.0.41"},{ssl,true}]},
      {load_definitions,none},
      {management_db_cache_multiplier,5},
      {max_http_body_size,10000000},
      {process_stats_gc_timeout,300000},
      {sample_retention_policies,
          [{global,[{86400,12000},{3600,600},{60,30}]},
           {basic,[{3600,600},{60,30}]},
           {detailed,[{10,5}]}]},
      {stats_event_max_backlog,250}]},
 {rabbitmq_management_agent,
     [{rates_mode,basic},
      {sample_retention_policies,
          [{global,[{605,5},{3660,60},{29400,600},{86400,1800}]},
           {basic,[{605,5},{3600,60}]},
           {detailed,[{605,5}]}]}]},
```
However, I found that rabbitmq_management_agent has some default values. I tried to set the values in the document and found that rabbitmq_management has the values I set. Should I change the values in rabbitmq_management_agent?

**Reply by michaelklishin** [reply]
@dormanze most configuration settings have default value. Without knowing which one you are talking about, we cannot suggest much. 

Are you sure you want to use the `advanced.config` file and mess with `rabbitmq_management_agent`? With `rabbitmq.conf` that's usually not necessary.

**Reply by dormanze** [reply]
My ultimate goal is to find a way to save memory resources as much as possible without closing the rest API. Because we found that the memory usage was increasing when the number of queues and channels remained unchanged and there was no message backlog for more than 15 days. I checked the memory details and found that the plug-in occupied a lot of memory. I don't know if this is a normal situation.
This is just after the restart:
![image](https://github.com/rabbitmq/rabbitmq-server/assets/37854724/13e28b8c-774a-4922-b96a-f008d17497f7)
The environment I asked above is what happened after 15 days of continuous operation.
Add a question. After the Prometheus plug-in is enabled, rabbitmq_management_agent and rabbitmq_web_dispatch are also enabled. Does it mean that metric data will also be cached in RabbitMQ?

**Reply by mkuratczyk** [reply]
`rabbitmq_management_agent` and `rabbitmq_prometheus` are independent of each other, so yes, enabling the Prometheus plugin doesn't automatically reduce the amount of data stored by the management plugin.

The best thing you can do is to lower the number of channels and queues you use. I don't see a good reason to have that many, if you have just ~300 connections to consume them.

As for the API - we recently made improvements to `/api/queues`, mostly focused on response time, but I guess this could affect memory usage as well. This particular change will be included in 3.13. Can you share what API endpoints you query and how (a full URI string with the filtering/pagination parameters)? It may or may not be related to your memory usage but either way would help us understand how the API is used and how we can make it more performant in future versions.

**Reply by dormanze** [reply]
I'm very grateful for your answer.
The requests I often use are as follows:
```
/api/health/checks/local-alarms
/api/vhosts
/api/nodes
/api/queues?page=1
/api/overview
/api/nodes/rabbit@adc-rabbitmqservice-0
/api/nodes
/api/health/checks/port-listener/rabbit@adc-rabbitmqservice-0
```
I also often view some queue information on the management interface.
Does the community consider not caching metric information, but using each query, providing the latest information each time? Because I've found that in the worst case, the cached information for metrics can take up half of the memory used, which is hard to accept.

**Reply by mkuratczyk** [reply]
I'm not sure I understand the question but if I do - that's basically why we suggest using Prometheus instead. We officially deprecated the Management stats 2+ years ago: https://blog.rabbitmq.com/posts/2021/08/4.0-deprecation-announcements/#disable-metrics-delivery-via-the-management-api--ui. However, many users love the built-in nature of the Management plugins/stats and it is sufficient for small deployments, so it's a hard to remove feature. We've not decided what exactly we'll do about it - perhaps changing the defaults to lower retention / granularity by default or something like that for now. But for large deployments - just disable Management stats and use prometheus (or at least set a very short retention period for the management stats)

**Reply by dormanze** [reply]
Thank you again for your detailed answer.
I tried to switch to the prometheus plug, and sure enough, the memory usage was about the same as the management plug, but the advantage was that the response was much faster. I will continue to adjust the metric collection period and cache.

**Reply by dormanze** [reply]
I finally decided to use the parameter management_agent.disable_metrics_collector: https://www.rabbitmq.com/management.html#disable-stats and turn on the rabbitmq_prometheus plugin.
